[NamedOptimizer][2/N] Prepare the enablement of state_dict for FSDP

fduwjj opened the pull request:
1. Add param_group check logic and unit test
2. Remove unnecessary check for conditional param update
3. Return the param_group from the inner optimizer so that when param_group is None or not all params are specified, we still return the expected result. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91147

Commit:
[NamedOptimizer][2/N] Enable state_dict for FSDP

[ghstack-poisoned]

fduwjj commented:
@pytorchbot rebase

Commit:
Update on "[NamedOptimizer][2/N] Prepare the enablement of state_dict for FSDP"

1. Add param_group check logic and unit test
2. Remove unnecessary check for conditional param update
3. Return the param_group from the inner optimizer so that when param_group is None or not all params are specified, we still return the expected result. 




[ghstack-poisoned]

fegin commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -171,9 +190,7 @@ def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
             )
         for idx, param_key in enumerate(self.ordered_param_keys):
             if param_key not in state.keys():
-                raise ValueError(
```
Can we add a comment to describe why this is legit?

fduwjj replied:
Sure. Let me explain it with comments.

fegin reviewed (approved):
Left a comment. Accept to unblock.

Commit:
Update on "[NamedOptimizer][2/N] Prepare the enablement of state_dict for FSDP"

1. Add param_group check logic and unit test
2. Remove unnecessary check for conditional param update
3. Return the param_group from the inner optimizer so that when param_group is None or not all params are specified, we still return the expected result. 




[ghstack-poisoned]

fduwjj commented:
@pytochbot merge

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot rebase

Commit:
Update on "[NamedOptimizer][2/N] Prepare the enablement of state_dict for FSDP"

1. Add param_group check logic and unit test
2. Remove unnecessary check for conditional param update
3. Return the param_group from the inner optimizer so that when param_group is None or not all params are specified, we still return the expected result. 




[ghstack-poisoned]

fduwjj commented:
@pytochbot merge

fduwjj commented:
@pytorchbot merge

rohan-varma commented on torch/distributed/optim/named_optimizer.py:
```diff
@@ -170,10 +189,9 @@ def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
                 f"Expects equal length as {len(new_state)} in `state_dict` state length but found {len(state)}."
             )
         for idx, param_key in enumerate(self.ordered_param_keys):
+            # When the conditional training is performed, not all parameters are updated in the optim.
```
Can we add a parity test with regular optimizer under conditional training use case?

fduwjj replied:
Done, added in https://github.com/pytorch/pytorch/pull/91344
